HIP/hip-415.md

```diff
@@ -0,0 +1,255 @@
+	}
+	if (`newBlock`) {
+			`blockHashes[blockNumber] = keccak256(runningHash)` // `runningHash` is stored in `RecordStreaming`. It is a running hash of the last processed RSO
+			delete `blockHashes[blockNumber - 256]`
+			`blockNumber++`
+			`blockTimestamp = currentTS`
+	}
+	processTransaction(tx)
+	...
+}
+
+Migration:
+`blockNumber = bootstrapLastBlockNumber`, where `bootstrapLastBlockNumber` is a system property added on startup.
+`blockTimestamp = bootstrapLastBlockTimestamp`, where `bootstrapLastBlockTimestamp` is a system property added on startup.
+```
+
+`number`, `timestamp` and `hash` (map of the 256 most recent blocks) must be available during transaction execution since the following opcodes are to be supported as per the EVM specification:
+
+- `BLOCKHASH` → Accepts the block `NUMBER` for which to return the `hash`. Valid range is the last `256` blocks (not including the current one)
+- `NUMBER` → Returns the current block number
+- `TIMESTAMP` → Returns the unix timestamp of the current block
+
+**Record File**
+
+Once HIP-435 is implemented, record files are to be migrated to protobufs. New `long blockNumber` property is to be 
+added in the record stream.
+
+It is required for `services` to propagate this property to `mirror nodes` since there are `partial mirror nodes`,
+that don’t keep the full history from the `first record file`. Due to that, they are unable to calculate the
+`block number`, thus all other `block properties`. Block `hash` and `timestamp` will not be included in the record files,
+since `block hash` is the `32 byte` prefix (out of `48 bytes`) of the running hash of the last `Record Stream 
+Object` from the previous `Record File` and `timestamp` is `1st Record Stream Object`'s consensus `timestamp`.
+
+With the introduction of the new version of `Record Stream Object`s, the respective libraries for state proofs must be updated:
+- [https://github.com/hashgraph/hedera-mirror-node/tree/main/hedera-mirror-rest/check-state-proof](https://github.com/hashgraph/hedera-mirror-node/tree/main/hedera-mirror-rest/check-state-proof)
+- [https://github.com/hashgraph/hedera-state-proof-verifier-go](https://github.com/hashgraph/hedera-state-proof-verifier-go)
+
+### Mirror Nodes
+
+Based on the updates specified above, the record stream objects will pass all the necessary information for Mirror Nodes to build `Record files`/`blocks` and store enough data about them to be able to answer all block related queries outlined above. 
+To do this they will need to read and store the `number` specified in the `Record file`. For 
+old record files, they will be derived through a migration process as specified below.
+
+**Historical State / Migration**
+
+`Record files` prior to the introduction of the block properties will not expose the required information. `mirror 
+nodes` that have full sync on `testnet` / `mainnet` networks will export a file containing mapping of record file 
```

Comment:
nit: We're not exporting it but just maintaining a copy in the mirror node's classpath.

Reply:
updated